server/src/main/java/org/elasticsearch/search/query/TopDocsCollectorContext.java

```diff
@@ -237,6 +235,13 @@ private SimpleTopDocsCollectorContext(IndexReader reader,
             super(REASON_SEARCH_TOP_HITS, numHits);
             this.sortAndFormats = sortAndFormats;
 
+            // disable max score optimization since we have a mandatory clause
+            // that doesn't track the maximum score
+            if ((sortAndFormats == null || SortField.FIELD_SCORE.equals(sortAndFormats.sort.getSort()[0]))
+                    && hasInfMaxScore(query)) {
+                trackTotalHitsUpTo = Integer.MAX_VALUE;
+            }
+
```

Comment:
We ignore `trackTotalHitsUpTo` [below](https://github.com/elastic/elasticsearch/pull/46105/files#diff-955b0a9153be35215d8032f9ca753100R260) if the hit count can be guessed from the query (`shortcutTotalHitCount`). We should not do that (ignore trackTotalHitsUpTo) if `hasInfMaxScore` is true.